File: services/api/src/security/password.js
const bcrypt = require("bcrypt");

function resolveSaltRounds() {
  const parsedRounds = Number(process.env.BCRYPT_SALT_ROUNDS || 12);
  if (!Number.isInteger(parsedRounds) || parsedRounds < 10) {
    return 12;
  }
  return parsedRounds;
}

async function hashPassword(plainTextPassword) {
  return bcrypt.hash(String(plainTextPassword), resolveSaltRounds());
}

async function verifyPassword(plainTextPassword, hashedPassword) {
  return bcrypt.compare(String(plainTextPassword), String(hashedPassword));
}

module.exports = {
  hashPassword,
  verifyPassword
};

File: services/api-python/app/security/password.py
import bcrypt

from app.config import get_settings

BCRYPT_MAX_PASSWORD_BYTES = 72


def resolve_salt_rounds() -> int:
    return get_settings().resolved_bcrypt_salt_rounds()


def hash_password(plain_text_password: str) -> str:
    password_bytes = str(plain_text_password).encode("utf-8")
    hashed = bcrypt.hashpw(password_bytes, bcrypt.gensalt(rounds=resolve_salt_rounds()))
    return hashed.decode("utf-8")


def verify_password(plain_text_password: str, hashed_password: str) -> bool:
    return bcrypt.checkpw(
        str(plain_text_password).encode("utf-8"),
        str(hashed_password).encode("utf-8"),
    )


def is_password_within_bcrypt_limit(password: str) -> bool:
    return len(password.encode("utf-8")) <= BCRYPT_MAX_PASSWORD_BYTES
